include/exec/trampoline_scheduler.hpp

```diff
@@ -0,0 +1,186 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ * Copyright (c) 2023 NVIDIA Corporation
+ *
+ * Licensed under the Apache License Version 2.0 with LLVM Exceptions
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *   https://llvm.org/LICENSE.txt
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include "../stdexec/execution.hpp"
+
+#include <cstddef>
+#include <type_traits>
+#include <utility>
+
+namespace exec {
+  namespace __trampoline {
+    using namespace stdexec;
+
+    template <class _Operation>
+    struct __trampoline_state {
+      static thread_local __trampoline_state* __current_;
+
+      __trampoline_state() noexcept {
+        __current_ = this;
+      }
+
+      ~__trampoline_state() {
+        __current_ = nullptr;
+      }
+
+      void __drain() noexcept;
+
+      std::size_t __recursion_depth_ = 1;
+      _Operation* __head_ = nullptr;
+    };
+
+    class __scheduler {
+      std::size_t __max_recursion_depth_;
+
+     public:
+      __scheduler() noexcept
+        : __max_recursion_depth_(16) {
+      }
+
+      explicit __scheduler(std::size_t __max_recursion_depth) noexcept
+        : __max_recursion_depth_(__max_recursion_depth) {
+      }
+
+     private:
+      struct __operation_base {
+        using __execute_fn = void(__operation_base*) noexcept;
+
+        explicit __operation_base(__execute_fn* __execute, std::size_t __max_depth) noexcept
+          : __execute_(__execute)
+          , __max_recursion_depth_(__max_depth) {
+        }
+
+        void __execute() noexcept {
+          __execute_(this);
+        }
+
+        friend void tag_invoke(start_t, __operation_base& __self) noexcept {
+          auto* __current_state = __trampoline_state<__operation_base>::__current_;
+          if (__current_state == nullptr) {
+            __trampoline_state<__operation_base> __state;
+            __self.__execute();
+            __state.__drain();
+          } else if (__current_state->__recursion_depth_ < __self.__max_recursion_depth_) {
+            ++__current_state->__recursion_depth_;
+            __self.__execute();
+          } else {
+            // Exceeded recursion limit.
+            __self.__next_ = std::exchange(
+              __current_state->__head_, static_cast<__operation_base*>(&__self));
+          }
```

Comment:
I admit I have shamelessly copied this implementation from libunifex (and left the Facebook copyright!). Now that I'm looking at this `start` implementation, I have Questions.

If we call `start` on one of these operation states and we've hit the recursion limit, we stuff the operation into the head of a thread-local list. The next time a trampoline operation is started on this thread, we execute all the operations in the list. Fine.

But what if we _don't_ start another trampoline operation on this thread? What becomes of the started operations sitting in the list? Program correctness could demand that this operations complete. If they don't, they can hang the program.

Am I confused?

Reply:
attn @kirkshoop @lewissbaker

Reply:
Maybe when the recursion limit is hit, we should post the work to the delegatee scheduler? And hard-error if there isn't one?

Reply:
When the recursion limit is reached it posts work back to the first call to start (the first if-branch where it calls drain()).

Each time execution returns to drain() loop it resets the recursion count back to 1 before executing the next item in the queue.